skipchain/skipchain.go

```diff
@@ -144,331 +173,427 @@ func (s *Service) ProposeSkipBlock(psbd *ProposeSkipBlock) (network.Message, one
+		log.Error("Didn't receive GetBlock")
+		return
+	}
+	sb := s.Sbm.GetByID(gb.ID)
+	if sb == nil {
+		log.Error("Did not find block")
 		return
 	}
-	if err := sb.VerifySignatures(); err != nil {
+	if i, _ := sb.Roster.Search(s.ServerIdentity().ID); i < 0 {
+		log.Lvl3("Not responsible for that block, recursing")
+		var err error
+		sb, err = s.getUpdateBlock(sb, sb.Hash)
+		if err != nil {
+			log.Error(err)
+			sb = nil
+		}
+	}
+	if err := s.SendRaw(env.ServerIdentity, &GetBlockReply{sb}); err != nil {
 		log.Error(err)
+	}
+}
+
+func (s *Service) getBlockReply(env *network.Envelope) {
+	gbr, ok := env.Msg.(*GetBlockReply)
+	if !ok {
+		log.Error("Didn't receive GetBlock")
 		return
 	}
-	s.storeSkipBlock(sb)
-	log.Lvlf3("Stored skip block %+v in %x", *sb, s.Context.ServerIdentity().ID[0:8])
+	if err := s.Sbm.VerifyLinks(gbr.SkipBlock); err != nil {
+		log.Error("Received invalid skipblock: " + err.Error())
+	}
+	id := s.Sbm.Store(gbr.SkipBlock)
+	s.save()
+	log.Lvl3("Sending block to channel")
+	s.blockRequests[string(id)] <- gbr.SkipBlock
+}
+
+// verifyFollowBlock makes sure that a signature-request for a forward-link
+// is valid.
+func (s *Service) bftVerifyFollowBlock(msg []byte, data []byte) bool {
+	err := func() error {
+		_, fsInt, err := network.Unmarshal(data)
+		if err != nil {
+			return err
+		}
+		fs, ok := fsInt.(*ForwardSignature)
+		if !ok {
+			return errors.New("Didn't receive a ForwardSignature")
+		}
+		previous := s.Sbm.GetByID(fs.Previous)
+		if previous == nil {
+			return errors.New("Didn't find newest block")
+		}
+		newest := fs.Newest
+		if len(newest.BackLinkIDs) <= fs.TargetHeight {
+			return errors.New("Asked for signing too high a backlink")
+		}
+		if err := fs.ForwardLink.VerifySignature(previous.Roster.Publics()); err != nil {
+			return errors.New("Wrong forward-link signature: " + err.Error())
+		}
+		if !fs.ForwardLink.Hash.Equal(newest.Hash) {
+			return errors.New("No forward-link from previous to newest")
+		}
+		target := s.Sbm.GetByID(newest.BackLinkIDs[fs.TargetHeight])
+		s.Sbm.Lock()
+		defer s.Sbm.Unlock()
+		if target == nil {
+			return errors.New("Don't have target-block")
+		}
+		if target.GetForwardLen() >= fs.TargetHeight+1 {
+			return errors.New("Already have forward-link at height " +
+				strconv.Itoa(fs.TargetHeight+1))
+		}
+		if !target.SkipChainID().Equal(newest.SkipChainID()) {
+			return errors.New("Target and newest not from same skipchain")
+		}
+		return nil
+	}()
+	if err != nil {
+		log.Error(err)
+		return false
+	}
+	return true
 }
 
-// VerifyShardFunc makes sure that the cothority of the child-skipchain is
-// part of the root-cothority.
-func (s *Service) VerifyShardFunc(msg []byte, sb *SkipBlock) bool {
-	if sb.ParentBlockID.IsNull() {
-		log.Lvl3("No parent skipblock to verify against")
+// verifyNewBlock makes sure that a signature-request for a forward-link
+// is valid.
+func (s *Service) bftVerifyNewBlock(msg []byte, data []byte) bool {
```

Comment:
Both are needed for the moment. I opened an issue for the v2-skipchains:

https://github.com/dedis/cothority/issues/870